<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><title>R: Phylogenetic analysis in R</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<link rel="stylesheet" type="text/css" href="R.css">
</head><body>
<h1> Phylogenetic analysis in R
<img class="toplogo" src="../../../doc/html/logo.jpg" alt="[R logo]">
</h1>
<hr>
<div align="center">
<a href="../../../doc/html/packages.html"><img src="../../../doc/html/left.jpg" alt="[Up]" width="30" height="30" border="0"></a>
<a href="../../../doc/html/index.html"><img src="../../../doc/html/up.jpg" alt="[Top]" width="30" height="30" border="0"></a>
</div><h2>Documentation for package &lsquo;phangorn&rsquo; version 1.99-11</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
<li><a href="../doc/index.html">User guides, package vignettes and other documentation.</a></li>
<li><a href="../NEWS">Package NEWS</a>.</li>
</ul>

<h2>Help Pages</h2>


<p align="center">
<a href="# "> </a>
<a href="#A">A</a>
<a href="#B">B</a>
<a href="#C">C</a>
<a href="#D">D</a>
<a href="#E">E</a>
<a href="#F">F</a>
<a href="#G">G</a>
<a href="#H">H</a>
<a href="#L">L</a>
<a href="#M">M</a>
<a href="#N">N</a>
<a href="#O">O</a>
<a href="#P">P</a>
<a href="#R">R</a>
<a href="#S">S</a>
<a href="#T">T</a>
<a href="#U">U</a>
<a href="#W">W</a>
<a href="#Y">Y</a>
</p>

<table width="100%">
<tr><td width="25%"><a href="phangorn-package.html">phangorn-package</a></td>
<td>Phylogenetic analysis in R</td></tr>
</table>

<h2><a name="A">-- A --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="parsimony.html">acctran</a></td>
<td>Parsimony tree.</td></tr>
<tr><td width="25%"><a href="phyDat.html">acgt2ry</a></td>
<td>Conversion among Sequence Formats</td></tr>
<tr><td width="25%"><a href="as.splits.html">addConfidences</a></td>
<td>Splits representation of graphs and trees.</td></tr>
<tr><td width="25%"><a href="phyDat.html">allSitePattern</a></td>
<td>Conversion among Sequence Formats</td></tr>
<tr><td width="25%"><a href="as.splits.html">allSplits</a></td>
<td>Splits representation of graphs and trees.</td></tr>
<tr><td width="25%"><a href="allTrees.html">allTrees</a></td>
<td>Compute all trees topologies.</td></tr>
<tr><td width="25%"><a href="Ancestors.html">Ancestors</a></td>
<td>tree utility function</td></tr>
<tr><td width="25%"><a href="ancestral.pml.html">ancestral.pars</a></td>
<td>Ancestral character reconstruction.</td></tr>
<tr><td width="25%"><a href="ancestral.pml.html">ancestral.pml</a></td>
<td>Ancestral character reconstruction.</td></tr>
<tr><td width="25%"><a href="phyDat.html">as.character.phyDat</a></td>
<td>Conversion among Sequence Formats</td></tr>
<tr><td width="25%"><a href="phyDat.html">as.data.frame.phyDat</a></td>
<td>Conversion among Sequence Formats</td></tr>
<tr><td width="25%"><a href="phyDat.html">as.DNAbin.phyDat</a></td>
<td>Conversion among Sequence Formats</td></tr>
<tr><td width="25%"><a href="as.splits.html">as.Matrix</a></td>
<td>Splits representation of graphs and trees.</td></tr>
<tr><td width="25%"><a href="as.splits.html">as.Matrix.splits</a></td>
<td>Splits representation of graphs and trees.</td></tr>
<tr><td width="25%"><a href="as.splits.html">as.matrix.splits</a></td>
<td>Splits representation of graphs and trees.</td></tr>
<tr><td width="25%"><a href="plot.networx.html">as.networx</a></td>
<td>Phylogenetic networks</td></tr>
<tr><td width="25%"><a href="plot.networx.html">as.networx.splits</a></td>
<td>Phylogenetic networks</td></tr>
<tr><td width="25%"><a href="phyDat.html">as.phyDat</a></td>
<td>Conversion among Sequence Formats</td></tr>
<tr><td width="25%"><a href="phyDat.html">as.phyDat.DNAbin</a></td>
<td>Conversion among Sequence Formats</td></tr>
<tr><td width="25%"><a href="as.splits.html">as.phylo.splits</a></td>
<td>Splits representation of graphs and trees.</td></tr>
<tr><td width="25%"><a href="as.splits.html">as.prop.part.splits</a></td>
<td>Splits representation of graphs and trees.</td></tr>
<tr><td width="25%"><a href="as.splits.html">as.splits</a></td>
<td>Splits representation of graphs and trees.</td></tr>
<tr><td width="25%"><a href="as.splits.html">as.splits.multiPhylo</a></td>
<td>Splits representation of graphs and trees.</td></tr>
<tr><td width="25%"><a href="as.splits.html">as.splits.networx</a></td>
<td>Splits representation of graphs and trees.</td></tr>
<tr><td width="25%"><a href="as.splits.html">as.splits.phylo</a></td>
<td>Splits representation of graphs and trees.</td></tr>
</table>

<h2><a name="B">-- B --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="bab.html">bab</a></td>
<td>Branch and bound for finding all most parsimonious trees</td></tr>
<tr><td width="25%"><a href="phyDat.html">baseFreq</a></td>
<td>Conversion among Sequence Formats</td></tr>
<tr><td width="25%"><a href="bootstrap.pml.html">bootstrap.phyDat</a></td>
<td>Bootstrap</td></tr>
<tr><td width="25%"><a href="bootstrap.pml.html">bootstrap.pml</a></td>
<td>Bootstrap</td></tr>
<tr><td width="25%"><a href="bab.html">BranchAndBound</a></td>
<td>Branch and bound for finding all most parsimonious trees</td></tr>
</table>

<h2><a name="C">-- C --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="Ancestors.html">Children</a></td>
<td>tree utility function</td></tr>
<tr><td width="25%"><a href="chloroplast.html">chloroplast</a></td>
<td>Chloroplast alignment</td></tr>
<tr><td width="25%"><a href="parsimony.html">CI</a></td>
<td>Parsimony tree.</td></tr>
<tr><td width="25%"><a href="cladePar.html">cladePar</a></td>
<td>Utility function to plot.phylo</td></tr>
<tr><td width="25%"><a href="superTree.html">coalSpeciesTree</a></td>
<td>Super Tree and Species Tree methods</td></tr>
<tr><td width="25%"><a href="as.splits.html">compatible</a></td>
<td>Splits representation of graphs and trees.</td></tr>
<tr><td width="25%"><a href="consensusNet.html">consensusNet</a></td>
<td>Computes a conensusNetwork from a list of trees Computes a 'networx' object from a collection of splits.</td></tr>
<tr><td width="25%"><a href="as.splits.html">countCycles</a></td>
<td>Splits representation of graphs and trees.</td></tr>
</table>

<h2><a name="D">-- D --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="densiTree.html">densiTree</a></td>
<td>Plots a densiTree.</td></tr>
<tr><td width="25%"><a href="Ancestors.html">Descendants</a></td>
<td>tree utility function</td></tr>
<tr><td width="25%"><a href="designTree.html">designSplits</a></td>
<td>Compute a design matrix or non-negative LS</td></tr>
<tr><td width="25%"><a href="designTree.html">designTree</a></td>
<td>Compute a design matrix or non-negative LS</td></tr>
<tr><td width="25%"><a href="dfactorial.html">dfactorial</a></td>
<td>Arithmetic Operators</td></tr>
<tr><td width="25%"><a href="dist.hamming.html">dist.hamming</a></td>
<td>Pairwise Distances from Sequences</td></tr>
<tr><td width="25%"><a href="dist.hamming.html">dist.logDet</a></td>
<td>Pairwise Distances from Sequences</td></tr>
<tr><td width="25%"><a href="dist.hamming.html">dist.ml</a></td>
<td>Pairwise Distances from Sequences</td></tr>
<tr><td width="25%"><a href="dist.p.html">dist.p</a></td>
<td>Pairwise Polymorphism P-Distances from DNA Sequences</td></tr>
<tr><td width="25%"><a href="distanceHadamard.html">distanceHadamard</a></td>
<td>Distance Hadamard</td></tr>
<tr><td width="25%"><a href="getClans.html">diversity</a></td>
<td>Clans, slices and clips</td></tr>
</table>

<h2><a name="E">-- E --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="pml.fit.html">edQt</a></td>
<td>Internal maximum likelihood functions.</td></tr>
</table>

<h2><a name="F">-- F --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="hadamard.html">fhm</a></td>
<td>Hadamard Matrices and Fast Hadamard Multiplication</td></tr>
<tr><td width="25%"><a href="parsimony.html">fitch</a></td>
<td>Parsimony tree.</td></tr>
</table>

<h2><a name="G">-- G --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="getClans.html">getClans</a></td>
<td>Clans, slices and clips</td></tr>
<tr><td width="25%"><a href="getClans.html">getClips</a></td>
<td>Clans, slices and clips</td></tr>
<tr><td width="25%"><a href="getClans.html">getDiversity</a></td>
<td>Clans, slices and clips</td></tr>
<tr><td width="25%"><a href="midpoint.html">getRoot</a></td>
<td>Tree manipulation</td></tr>
<tr><td width="25%"><a href="getClans.html">getSlices</a></td>
<td>Clans, slices and clips</td></tr>
</table>

<h2><a name="H">-- H --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="hadamard.html">h2st</a></td>
<td>Hadamard Matrices and Fast Hadamard Multiplication</td></tr>
<tr><td width="25%"><a href="hadamard.html">h4st</a></td>
<td>Hadamard Matrices and Fast Hadamard Multiplication</td></tr>
<tr><td width="25%"><a href="hadamard.html">hadamard</a></td>
<td>Hadamard Matrices and Fast Hadamard Multiplication</td></tr>
</table>

<h2><a name="L">-- L --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="Laurasiatherian.html">Laurasiatherian</a></td>
<td>Laurasiatherian data (AWCMEE)</td></tr>
<tr><td width="25%"><a href="dfactorial.html">ldfactorial</a></td>
<td>Arithmetic Operators</td></tr>
<tr><td width="25%"><a href="lento.html">lento</a></td>
<td>Lento plot</td></tr>
<tr><td width="25%"><a href="pml.fit.html">lli</a></td>
<td>Internal maximum likelihood functions.</td></tr>
</table>

<h2><a name="M">-- M --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="midpoint.html">midpoint</a></td>
<td>Tree manipulation</td></tr>
<tr><td width="25%"><a href="modelTest.html">modelTest</a></td>
<td>ModelTest</td></tr>
<tr><td width="25%"><a href="Ancestors.html">mrca.phylo</a></td>
<td>tree utility function</td></tr>
</table>

<h2><a name="N">-- N --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="NJ.html">NJ</a></td>
<td>Neighbor-Joining</td></tr>
<tr><td width="25%"><a href="nni.html">nni</a></td>
<td>Tree rearrangements.</td></tr>
<tr><td width="25%"><a href="designTree.html">nnls.tree</a></td>
<td>Compute a design matrix or non-negative LS</td></tr>
</table>

<h2><a name="O">-- O --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="parsimony.html">optim.parsimony</a></td>
<td>Parsimony tree.</td></tr>
<tr><td width="25%"><a href="pml.html">optim.pml</a></td>
<td>Likelihood of a tree.</td></tr>
</table>

<h2><a name="P">-- P --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="ancestral.pml.html">pace</a></td>
<td>Ancestral character reconstruction.</td></tr>
<tr><td width="25%"><a href="parsimony.html">parsimony</a></td>
<td>Parsimony tree.</td></tr>
<tr><td width="25%"><a href="phangorn-package.html">phangorn</a></td>
<td>Phylogenetic analysis in R</td></tr>
<tr><td width="25%"><a href="phyDat.html">phyDat</a></td>
<td>Conversion among Sequence Formats</td></tr>
<tr><td width="25%"><a href="plot.networx.html">plot.networx</a></td>
<td>Phylogenetic networks</td></tr>
<tr><td width="25%"><a href="ancestral.pml.html">plotAnc</a></td>
<td>Ancestral character reconstruction.</td></tr>
<tr><td width="25%"><a href="bootstrap.pml.html">plotBS</a></td>
<td>Bootstrap</td></tr>
<tr><td width="25%"><a href="pml.html">pml</a></td>
<td>Likelihood of a tree.</td></tr>
<tr><td width="25%"><a href="pml.html">pml.control</a></td>
<td>Likelihood of a tree.</td></tr>
<tr><td width="25%"><a href="pml.fit.html">pml.fit</a></td>
<td>Internal maximum likelihood functions.</td></tr>
<tr><td width="25%"><a href="pml.fit.html">pml.free</a></td>
<td>Internal maximum likelihood functions.</td></tr>
<tr><td width="25%"><a href="pml.fit.html">pml.init</a></td>
<td>Internal maximum likelihood functions.</td></tr>
<tr><td width="25%"><a href="pmlCluster.html">pmlCluster</a></td>
<td>Stochastic Partitioning</td></tr>
<tr><td width="25%"><a href="pmlMix.html">pmlMix</a></td>
<td>Phylogenetic mixture model</td></tr>
<tr><td width="25%"><a href="pmlPart.html">pmlPart</a></td>
<td>Partition model.</td></tr>
<tr><td width="25%"><a href="pmlMix.html">pmlPen</a></td>
<td>Phylogenetic mixture model</td></tr>
<tr><td width="25%"><a href="parsimony.html">PNJ</a></td>
<td>Parsimony tree.</td></tr>
<tr><td width="25%"><a href="parsimony.html">pratchet</a></td>
<td>Parsimony tree.</td></tr>
<tr><td width="25%"><a href="as.splits.html">presenceAbsence</a></td>
<td>Splits representation of graphs and trees.</td></tr>
<tr><td width="25%"><a href="as.splits.html">print.splits</a></td>
<td>Splits representation of graphs and trees.</td></tr>
<tr><td width="25%"><a href="midpoint.html">pruneTree</a></td>
<td>Tree manipulation</td></tr>
</table>

<h2><a name="R">-- R --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="parsimony.html">random.addition</a></td>
<td>Parsimony tree.</td></tr>
<tr><td width="25%"><a href="read.aa.html">read.aa</a></td>
<td>Read Amino Acid Sequences in a File</td></tr>
<tr><td width="25%"><a href="as.splits.html">read.nexus.splits</a></td>
<td>Splits representation of graphs and trees.</td></tr>
<tr><td width="25%"><a href="phyDat.html">read.phyDat</a></td>
<td>Conversion among Sequence Formats</td></tr>
<tr><td width="25%"><a href="treedist.html">RF.dist</a></td>
<td>Distances between trees</td></tr>
<tr><td width="25%"><a href="parsimony.html">RI</a></td>
<td>Parsimony tree.</td></tr>
<tr><td width="25%"><a href="nni.html">rNNI</a></td>
<td>Tree rearrangements.</td></tr>
<tr><td width="25%"><a href="nni.html">rSPR</a></td>
<td>Tree rearrangements.</td></tr>
</table>

<h2><a name="S">-- S --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="parsimony.html">sankoff</a></td>
<td>Parsimony tree.</td></tr>
<tr><td width="25%"><a href="SH.test.html">SH.test</a></td>
<td>Shimodaira-Hasegawa Test</td></tr>
<tr><td width="25%"><a href="Ancestors.html">Siblings</a></td>
<td>tree utility function</td></tr>
<tr><td width="25%"><a href="simSeq.html">simSeq</a></td>
<td>Simulate sequences.</td></tr>
<tr><td width="25%"><a href="simSeq.html">simSeq.phylo</a></td>
<td>Simulate sequences.</td></tr>
<tr><td width="25%"><a href="simSeq.html">simSeq.pml</a></td>
<td>Simulate sequences.</td></tr>
<tr><td width="25%"><a href="SOWH.test.html">SOWH.test</a></td>
<td>Swofford-Olsen-Waddell-Hillis Test</td></tr>
<tr><td width="25%"><a href="splitsNetwork.html">splitsNetwork</a></td>
<td>Phylogenetic Network</td></tr>
<tr><td width="25%"><a href="phyDat.html">subset.phyDat</a></td>
<td>Conversion among Sequence Formats</td></tr>
<tr><td width="25%"><a href="superTree.html">superTree</a></td>
<td>Super Tree and Species Tree methods</td></tr>
</table>

<h2><a name="T">-- T --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="treedist.html">treedist</a></td>
<td>Distances between trees</td></tr>
</table>

<h2><a name="U">-- U --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="NJ.html">UNJ</a></td>
<td>Neighbor-Joining</td></tr>
<tr><td width="25%"><a href="upgma.html">upgma</a></td>
<td>UPGMA and WPGMA</td></tr>
</table>

<h2><a name="W">-- W --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="upgma.html">wpgma</a></td>
<td>UPGMA and WPGMA</td></tr>
<tr><td width="25%"><a href="as.splits.html">write.nexus.splits</a></td>
<td>Splits representation of graphs and trees.</td></tr>
<tr><td width="25%"><a href="phyDat.html">write.phyDat</a></td>
<td>Conversion among Sequence Formats</td></tr>
<tr><td width="25%"><a href="as.splits.html">write.splits</a></td>
<td>Splits representation of graphs and trees.</td></tr>
</table>

<h2><a name="Y">-- Y --</a></h2>

<table width="100%">
<tr><td width="25%"><a href="yeast.html">yeast</a></td>
<td>Yeast alignment (Rokas et al.)</td></tr>
</table>
</body></html>
